### PR TITLE
Stricter pact parser

### PR DESCRIPTION
**Pull request by larskuhtz**
* [x] pact replay passed

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -36,7 +36,7 @@ debug-info: True
 source-repository-package
     type: git
     location: https://github.com/kadena-io/pact.git
-    tag: 51f8d6eeca1a8cff11d2fda5446522a4148f785c
+    tag: 21acdc69f80d08f8f576fcafe794ae360b178951
```
Comment (larskuhtz) [inline cabal.project:39]: This must be updated once the respective PR is merged into pact master.

File: dep/pact/github.json [hunk 2]
```diff
@@ -1,8 +1,8 @@
 {
   "owner": "kadena-io",
   "repo": "pact",
-  "branch": "rel/v4.1.2",
+  "branch": "rel/v4.2.0",
   "private": false,
-  "rev": "51f8d6eeca1a8cff11d2fda5446522a4148f785c",
-  "sha256": "0zrc720f86zz2zqbmcmc8qh84fwsx6hdmlwi0n9cdi38yz3ybkcq"
+  "rev": "21acdc69f80d08f8f576fcafe794ae360b178951",
+  "sha256": "0zrc720f86zz2zqbmcmc8qh84fwsx6hdmlwi0n9cdi38yz3ybkc0"
```
Comment (larskuhtz) [inline dep/pact/github.json:7]: The sha256 hash is yet invalid and must be replaced by the correct hash

File: src/Chainweb/Version.hs [hunk 3]
```diff
@@ -911,7 +912,13 @@ doCheckTxHash Mainnet01 = (>= 2_349_800) -- 2022-01-23T02:53:38
 doCheckTxHash Testnet04 = (>= 1_889_000) -- 2022-01-24T04:19:24
 doCheckTxHash Development = (>= 110)
 doCheckTxHash (FastTimedCPM g) | g == petersonChainGraph = (>= 7)
-doCheckTxHash _ = const False
+doCheckTxHash _ = const True
```
Comment (larskuhtz) [inline src/Chainweb/Version.hs:915]: Was there are reason for not checking the txhash in test chainweb versions?

**Review by edmundnoble — APPROVED** [review]

File: src/Chainweb/Transaction.hs [hunk 4]
```diff
@@ -91,26 +94,38 @@ instance Hashable (HashableTrans PayloadWithText) where
...
 
-parsePact :: Text -> Either String ParsedCode
-parsePact code = ParsedCode code <$> parseExprs code
+parsePact
+    :: Maybe (ChainwebVersion, BlockHeight)
+        -- ^ If the chain context is @Nothing@, latest parser version is used.
+    -> Text
+    -> Either String ParsedCode
+parsePact Nothing code = P.parsePact code
+parsePact (Just (v, h)) code
```
Comment (jmcardon) [inline src/Chainweb/Transaction.hs:126]: :+1:

**Review by jmcardon — APPROVED** [review]

**Pull request merged**